incubator/hnc/internal/pkg/selector/selector.go

```diff
@@ -0,0 +1,67 @@
+package selector
+
+import (
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+
+	api "sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha2"
+)
+
+func GetSelectorAnnotation(inst *unstructured.Unstructured) (string, bool) {
+	annot := inst.GetAnnotations()
+	str, ok := annot[api.AnnotationSelector]
+	return str, ok
+}
+
+// GetTreeSelector is similar to a regular selector, except that it adds the LabelTreeDepthSuffix to every string
+// To transform a tree selector into a regular label selector, we follow these steps:
+// 1. get the treeSelector annotation if it exists
+// 2. convert the annotation string to a slice of strings seperated by comma, because user is allowed to put multiple selectors
+// 3. append the LabelTreeDepthSuffix to each of the treeSelector string
+// 4. combine them into a single string connected by comma
+func GetTreeSelector(log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	annot := inst.GetAnnotations()
+	treeSelectorStr, ok := annot[api.AnnotationTreeSelector]
+	if !ok {
+		return nil
+	}
+	strs := strings.Split(treeSelectorStr, ",")
+	selectorStr := ""
+	for i, str := range strs {
+		selectorStr = selectorStr + str + api.LabelTreeDepthSuffix
+		if i < len(strs)-1 {
+			selectorStr = selectorStr + ", "
+		}
+	}
+	return GetSelectorFromString(log, selectorStr)
+}
+
+// GetSelector returns the selector on a given object if it exists
+func GetSelector(log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	selectorStr, ok := GetSelectorAnnotation(inst)
+	if !ok {
+		return nil
+	}
+	return GetSelectorFromString(log, selectorStr)
+}
+
+// getSelectorFromString converts the given string to a selector
```

Comment:
Why do we need to export this function? Why doesn't the validator just call `GetSelector` with `newInst`?

If we can't use `GetSelector`, I'd still like to hide this function because it seems like overkill to export it - I _think_ the only context it's used from outside of this package is to test whether a string is valid, not because we actually need the `labels.Selector` object. If that's correct, maybe we can export a `ValidateSelector` function instead, and leave this one as an unexported function.

Reply:
done